core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java

```diff
@@ -436,20 +442,103 @@ public long responseTimeoutMillis() {
         return responseTimeoutMillis;
     }
 
+    @Override
+    public void clearResponseTimeout() {
+        if (responseTimeoutMillis == 0) {
+            return;
+        }
+
+        final TimeoutController responseTimeoutController = this.responseTimeoutController;
+        responseTimeoutMillis = 0;
+        if (responseTimeoutController != null) {
+            if (eventLoop().inEventLoop()) {
+                responseTimeoutController.cancelTimeout();
+            } else {
+                eventLoop().execute(responseTimeoutController::cancelTimeout);
+            }
+        }
+    }
+
     @Override
     public void setResponseTimeoutMillis(long responseTimeoutMillis) {
-        if (responseTimeoutMillis < 0) {
-            throw new IllegalArgumentException(
-                    "responseTimeoutMillis: " + responseTimeoutMillis + " (expected: >= 0)");
+        checkArgument(responseTimeoutMillis >= 0,
+                      "responseTimeoutMillis: " + responseTimeoutMillis + " (expected: >= 0)");
+        if (responseTimeoutMillis == 0) {
+            clearResponseTimeout();
         }
-        this.responseTimeoutMillis = responseTimeoutMillis;
+
+        final long adjustmentMillis =
+                LongMath.saturatedSubtract(responseTimeoutMillis, this.responseTimeoutMillis);
+        extendResponseTimeoutMillis(adjustmentMillis);
     }
 
     @Override
     public void setResponseTimeout(Duration responseTimeout) {
         setResponseTimeoutMillis(requireNonNull(responseTimeout, "responseTimeout").toMillis());
     }
 
+    @Override
+    public void extendResponseTimeoutMillis(long adjustmentMillis) {
+        if (adjustmentMillis == 0 || responseTimeoutMillis == 0) {
+            return;
+        }
+
+        final long oldResponseTimeoutMillis = responseTimeoutMillis;
+        responseTimeoutMillis = LongMath.saturatedAdd(oldResponseTimeoutMillis, adjustmentMillis);
+        final TimeoutController responseTimeoutController = this.responseTimeoutController;
+        if (responseTimeoutController != null) {
+            if (eventLoop().inEventLoop()) {
+                responseTimeoutController.extendTimeout(adjustmentMillis);
+            } else {
+                eventLoop().execute(() -> responseTimeoutController.extendTimeout(adjustmentMillis));
+            }
+        }
+    }
+
+    @Override
+    public void extendResponseTimeout(Duration adjustment) {
+        extendResponseTimeoutMillis(requireNonNull(adjustment, "adjustment").toMillis());
+    }
+
+    @Override
+    public void setResponseTimeoutAfterMillis(long responseTimeoutMillis) {
+        checkArgument(responseTimeoutMillis > 0,
```

Comment:
What happens if `this.responseTimeoutMillis == 0`? I just wanted to make sure that we set the timeout when initial timeout is disabled.

Reply:
The timeout should be resumed and triggered after the `responseTimeoutMillis`.
I think there is something I missed. Let me take a look. 🙏